hbase-compression/hbase-compression-snappy/src/main/java/org/apache/hadoop/hbase/io/compress/xerial/SnappyCompressor.java

```diff
@@ -0,0 +1,159 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements. See the NOTICE file distributed with this
+ * work for additional information regarding copyright ownership. The ASF
+ * licenses this file to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.hadoop.hbase.io.compress.xerial;
+
+import java.io.IOException;
+import java.nio.ByteBuffer;
+
+import org.apache.hadoop.conf.Configuration;
+import org.apache.hadoop.io.compress.Compressor;
+import org.apache.yetus.audience.InterfaceAudience;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.xerial.snappy.Snappy;
+
+/**
+ * Hadoop compressor glue for Xerial Snappy.
+ */
+@InterfaceAudience.Private
+public class SnappyCompressor implements Compressor {
+
+  protected static final Logger LOG = LoggerFactory.getLogger(SnappyCompressor.class);
+
+  protected ByteBuffer inBuf;
+  protected ByteBuffer outBuf;
+  protected boolean finish, finished;
+  protected long bytesRead, bytesWritten;
+
+  SnappyCompressor(int bufferSize) {
+    this.inBuf = ByteBuffer.allocateDirect(bufferSize);
+    this.outBuf = ByteBuffer.allocateDirect(bufferSize);
+    this.outBuf.position(bufferSize);
+  }
+
+  @Override
+  public int compress(byte[] b, int off, int len) throws IOException {
```

Comment:
Unlike the other two wrapper implementations (for aircompressor and lz4) this one is slightly different because the Snappy API doesn't adjust the output buffer position like the others do. 

Although the code is similar, I am not sure we should abstract this to a common base class, because each upstream codec implementation can change over time.

Reply:
> Unlike the other two wrapper implementations (for aircompressor and lz4) this one is slightly different because the Snappy API doesn't adjust the output buffer position like the others do.

Even Zstd belongs to the same category as Snappy with this regard?